test/widgets/compose_box_test.dart

```diff
@@ -195,10 +206,163 @@ void main() {
+      await prepareComposeBox(tester, narrow: narrow, account: account);
+
+      await tester.enterText(topicInputFinder, destinationNarrow.topic);
+      // Remove an irrelevant topic request.
+      check(connection.takeRequests()).single
+        ..method.equals('GET')
+        ..url.path.equals('/api/v1/users/me/123/topics');
+
+      await checkStartTyping(tester, destinationNarrow);
+
+      connection.prepare(json: {});
+      await tester.pump(store.typingNotifier.typingStoppedWaitPeriod);
+      checkTypingRequest(TypingOp.stop, destinationNarrow);
+    });
+
+    testWidgets('clearing text stops typing notice', (tester) async {
+      await prepareComposeBox(tester, narrow: narrow, account: account);
+
+      await checkStartTyping(tester, narrow);
+
+      connection.prepare(json: {});
+      await tester.enterText(contentInputFinder, '');
+      checkTypingRequest(TypingOp.stop, narrow);
+    });
+
+    Future<void> prepareComposeBoxWithNavigation(WidgetTester tester) async {
+      addTearDown(testBinding.reset);
+      await testBinding.globalStore.add(account, eg.initialSnapshot(
+        zulipFeatureLevel: account.zulipFeatureLevel));
+
+      await tester.pumpWidget(const ZulipApp());
+      await tester.pump();
+      final navigator = await ZulipApp.navigator;
+      unawaited(navigator.push(MaterialAccountWidgetRoute(
+        accountId: account.id, page: const ComposeBox(narrow: narrow))));
+      await tester.pumpAndSettle();
+
+      store = await testBinding.globalStore.perAccount(account.id);
+      connection = store.connection as FakeApiConnection;
+    }
+
+    testWidgets('navigating away sends "stopped typing" notice', (tester) async {
+      await prepareComposeBoxWithNavigation(tester);
+
+      await checkStartTyping(tester, narrow);
+
+      connection.prepare(json: {});
+      (await ZulipApp.navigator).pop();
+      await tester.pump(Duration.zero);
+      checkTypingRequest(TypingOp.stop, narrow);
+    });
+
+    testWidgets('for content input, unfocusing sends "stopped typing" notice '
+                'and refocusing does not send "started typing notice', (tester) async {
+      const narrow = ChannelNarrow(123);
+      final destinationNarrow = TopicNarrow(narrow.streamId, 'test topic');
+      await prepareComposeBox(tester, narrow: narrow, account: account);
+
+      await tester.enterText(topicInputFinder, destinationNarrow.topic);
+      // Remove an irrelevant topic request.
+      check(connection.takeRequests()).single
+        ..method.equals('GET')
+        ..url.path.equals('/api/v1/users/me/123/topics');
+
+      connection.prepare(json: {});
+      await tester.enterText(contentInputFinder, 'hello world');
+      checkTypingRequest(TypingOp.start, destinationNarrow);
+
+      connection.prepare(json: {});
+      // Move focus to the topic input.
+      await tester.tap(topicInputFinder);
+      await tester.pump(Duration.zero);
+      checkTypingRequest(TypingOp.stop, destinationNarrow);
```

Comment:
Hmm, this actually seems a bit buggy. The user contemplates changing what topic they're sending to, and that causes us to decide they're not preparing a message after all?

If it's annoying to avoid this while getting the behavior we do want, then it's fine — *probably* the next thing the user is going to do is edit the topic, and then indeed we want to decide they're not sending to the old topic. But for the test, we should pick a scenario where the behavior is actually desirable.

I think the prime case for the "stop on unfocus" behavior is when the user touches the message list, or somewhere else outside the compose box, in order to unfocus the compose box and dismiss the keyboard.

---

… Oh and this also causes me to notice: on Android, once the focus is in the compose box it doesn't seem possible to get it out! At least, the cursor keeps blinking no matter what I do, short of navigating back from the whole page and then navigating into it again. This mostly doesn't affect the user — I think I hadn't previously noticed it in all this time using the app — but it does mean that this "stop on unfocus" behavior basically doesn't work on Android.

I think fixing that will be tough, though. I looked around and couldn't find an API for observing when the keyboard gets closed, or even checking whether it's open — the answers for that on the web are mostly just to look at the geometry of the window and try to guess.

So for this test, maybe just do this:
```dart
            FocusManager.instance.primaryFocus!.unfocus();
```

That's copied from the implementation in `ScrollView` of the semantics of `ScrollViewKeyboardDismissBehavior.onDrag`. Which is what we pass in `MessageList` to get the behavior we have on iOS where if you drag the message list it closes the keyboard. (The closing of the keyboard seems to come as a side effect of unfocusing the text input.)